### PR TITLE
feat(template): use environment variables in templates

### LINKED ISSUE #18765: Add `getEnvVariable` Handlebars helper
### What would you like Renovate to be able to do?

Renovate should be able to read environment variables, so they can be written to the PR body (e.g. through `prFooter` or `prHeader`).

This could be used, for example, to publish information from a GitLab job running `renovate`, e.g. `$CI_JOB_URL` (or other [predefined_variables](https://docs.gitlab.com/ee/ci/variables/predefined_variables.html)).

Related discussion: https://github.com/renovatebot/renovate/discussions/18763

### If you have any ideas on how this should be implemented, please tell us here.

The solution will be similar to https://github.com/renovatebot/renovate/pull/15212.

### Is this a feature you are interested in implementing yourself?

Yes

### PR DESCRIPTION
**Pull request by chopeen**
## Changes

I modified `template` module to load child process environment variables into list of allowed template fields. This allows to use selected environment variables anywhere where templating is used, e.g. to include additional information in `prHeader` or `prFooter`.

## Context

Closes #18765.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — CHANGES_REQUESTED** [review]

File: docs/usage/templates.md
Comment (chopeen) [reply docs/usage/templates.md:93]: 1. When spaces are added inside `{{ env.HOME }}`, regex `templateRegex` does not extract the variable name. Are you sure they should be there?
2. Prefix `env.` implemented.

File: lib/util/template/index.ts
Comment (chopeen) [reply lib/util/template/index.ts]: Prefix `env.` implemented.

File: lib/util/template/index.ts
Comment (chopeen) [inline lib/util/template/index.ts:191]: Are separate lists `allowedTemplateFields` and `allowedTemplateFields` necessary?

I would like to refactor this and keep only one of them (`allowedTemplateFields` because it has more fields?), but I think two different lists may be here on purpose.

File: lib/util/template/index.spec.ts
Comment (viceice) [inline lib/util/template/index.spec.ts:95]:
```suggestion
    const output = template.compile(userTemplate, {});
```

File: lib/util/template/index.spec.ts
Comment (viceice) [inline lib/util/template/index.spec.ts:106]:
```suggestion
    const output = template.compile(userTemplate, {});
```

File: lib/util/template/index.spec.ts
Comment (viceice) [inline lib/util/template/index.spec.ts:117]:
```suggestion
    const output = template.compile(userTemplate, {});
```

File: lib/util/template/index.spec.ts
Comment (viceice) [inline lib/util/template/index.spec.ts:123]:
```suggestion
    const output = template.compile(userTemplate, {});
```

File: lib/util/template/index.ts
Comment (viceice) [inline lib/util/template/index.ts:234]: move outside the loop

File: lib/util/template/index.ts
Comment (viceice) [inline lib/util/template/index.ts:165]: this is very bad for perf. we should refactor the code to pass from initial compile function

File: lib/util/template/index.ts
Comment (viceice) [reply lib/util/template/index.ts:191]: yes

File: lib/util/template/index.ts
Comment (viceice) [inline lib/util/template/index.ts:193]: create const in module scope for this (perf)

File: lib/util/template/index.ts
Comment (viceice) [reply lib/util/template/index.ts:234]: otherwise it will be evaluated for each match again, but that set never changes inside the loop

File: lib/util/template/index.ts
Comment (chopeen) [reply lib/util/template/index.ts]: I restored the original implementation with consts.

However, I now struggle with two unit tests I added - I don't know how to merge two `env` dictionaries with `const data = { ...GlobalConfig.get(), ...input, env: getChildEnv({}) };`. Probably my misunderstanding of the spread operator `...`. Can you advise please?

Comment (chopeen) [reply lib/util/template/index.ts]: You can disregard my question, I sorted it out. 

The two tests are still failing, but now I know why. I will post a new question when I'm at my desk.

File: lib/util/template/index.spec.ts
Comment (chopeen) [inline lib/util/template/index.spec.ts:97]: @viceice Here's the problem I am facing:

- this file imports the module `template`;
- consts `allowedTemplateFields` and `allowedFieldsList` get initialized when the module is loaded;
- no matter what I do in the unit test - even if I added `SHELL` to `GlobalConfig.customEnvVariables` - `SHELL` will not be added to the allowed fields.

How should I construct this test?

I can see two options:

1. In Python, I would override the consts values set in the module by monkeypatching them for the test. `mockReturnValueOnce` seems promising.
2. Reloading `template` after setting `env` and `customEnvVariables` would solve the issue.

Which is the proper way to go?

Comment (rarkins) [inline lib/util/template/index.spec.ts:97]:
```suggestion
        SHELL: '/bin/bash',
```

Can't this just a plain string?

**Comment by chopeen** [comment]
I fixed the Lodash `merge` call and now old unit tests are green again.

Two of my new ones are failing because of [19301#discussion_r1044543836](https://github.com/renovatebot/renovate/pull/19301#discussion_r1044543836) - can you advise please?

File: lib/util/template/index.ts
Comment (chopeen) [inline lib/util/template/index.ts:219]: What's the proper order here? Should `input` come last so it can override the global config and the child `env`?

File: lib/util/template/index.spec.ts
Comment (chopeen) [reply lib/util/template/index.spec.ts]: OK, I found a way. Not sure if it is the most elegant one, but the tests are :green_circle:.

**Review by viceice — CHANGES_REQUESTED** [review]
Still won't work as expected. I'll push some changes later, which work.

File: lib/util/template/index.ts
Comment (viceice) [inline lib/util/template/index.ts:219]:
```suggestion
  const data = { ...GlobalConfig.get(), ...input, env: getChildEnv({}) };
```

File: lib/util/template/index.ts
Comment (viceice) [inline lib/util/template/index.ts:184]:
```suggestion
    if (prop === 'env') {
      return target[prop];
    }
    if (!allowedTemplateFields.has(prop)) {
```

File: lib/util/template/index.ts
Comment (viceice) [inline lib/util/template/index.ts:180]: revert those changes

File: lib/util/template/index.ts
Comment (viceice) [inline lib/util/template/index.ts:223]:
```suggestion
    const matches = template.matchAll(templateRegex);
    const allowedFields = new Set([...allowedFieldsList, ...Object.keys(getChildEnv({}))]);
```

File: lib/util/template/index.ts
Comment (viceice) [inline lib/util/template/index.ts:227]:
```suggestion
        if (!allowedFields.has(varName)) {
```

File: package.json
Comment (viceice) [inline package.json:161]: revert, totally wrong

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/util/template/index.ts
Comment (chopeen) [reply lib/util/template/index.ts]: Thank you for the suggestions, @viceice! I implemented all of them.

However, that brings back a bug (that Lodash `merge` was solving):

```
> input
env: {CUSTOM_FOO: 'foo'}
> childEnv
{HOME: '/root', PATH: '/tmp/yarn--...', LC_ALL: 'C.UTF-8', LANG: 'C.UTF-8'}

> { ...input, env: childEnv }
env: {HOME: '/root', PATH: '/tmp/yarn--...', LC_ALL: 'C.UTF-8', LANG: 'C.UTF-8'}

> { env: childEnv, ...input }
env: {CUSTOM_FOO: 'foo'}
```

In the example above, both `input` and `childEnv` contain key `env`. Merging then with spread operators *keeps only one `env`* (last one).

Is there a built-in method to perform a deep (recursive) merge *or* should I implement one myself?

Comment (viceice) [reply lib/util/template/index.ts]: input shouldn't have env, as it was never supported. 🤔 why input has an env key?

lodash can't be used, as it's not a production dependency (moving type dep doesn't help) and we don't want to have that.

Comment (chopeen) [reply lib/util/template/index.ts]: The new unit tests I added use `input` to simulate custom variables that in production would be defined in `customEnvVariables` and arrive as `{ env: getChildEnv({}) }` in function `compile`.

Anyway, I found `deepmerge` used elsewhere and merged all variables this way. All tests are :green_circle: again.

File: lib/util/template/index.ts
Comment (viceice) [inline lib/util/template/index.ts:214]: don't merge! fix tests instead. input should never have an `env` key

File: lib/util/template/index.ts
Comment (viceice) [reply lib/util/template/index.ts]: see other conmment. don't merge, don't pass env via input. simply mock `getChildEnv`, so it returns the expected values

File: lib/util/template/index.ts
Comment (chopeen) [reply lib/util/template/index.ts]: Mocking! 🤦

Done.

---

I'm not 100% happy about my implementation, because I don't know how to mock `getChildEnv` selectively, so that some tests use the original implementation and some the mocked one.

I removed those of my new unit tests that required the original `getChildEnv`.

File: docs/usage/templates.md
Comment (viceice) [inline docs/usage/templates.md:91]: We should mention `exposeAllEnv`[^1] here to allow all env. @HonkingGoose Can you help?

[^1]: https://docs.renovatebot.com/self-hosted-configuration/#exposeallenv

File: lib/util/template/index.spec.ts
Comment (viceice) [inline lib/util/template/index.spec.ts:21]:
```suggestion
```
shouldn't needed

File: lib/util/template/index.ts
Comment (viceice) [inline lib/util/template/index.ts:209]:
```suggestion
  const env = getChildEnv({})
  const data = { ...GlobalConfig.get(), ...input, env) };
```

File: lib/util/template/index.ts
Comment (viceice) [inline lib/util/template/index.ts:216]:
```suggestion
      ...Object.keys(env),
```

**Review by viceice — COMMENTED** [review]
some small improvements

**Review by viceice — DISMISSED** [review]

File: docs/usage/templates.md
Comment (HonkingGoose) [inline docs/usage/templates.md:98]:
```suggestion
You can also use the [`exposeAllEnv`](https://docs.renovatebot.com/self-hosted-configuration/#exposeallenv) config option to allow all environment variables in templates.
Always consider the security implications of using `exposeAllEnv`.
```

File: docs/usage/templates.md
Comment (HonkingGoose) [inline docs/usage/templates.md:100]:
```suggestion
Secrets and other confidential information stored in environment variables could be leaked by a malicious script.
```

File: docs/usage/templates.md
Comment (HonkingGoose) [inline docs/usage/templates.md:91]:
```suggestion
By default, you can only access a handful of basic environment variables like `HOME` or `PATH`.
This is for security reasons.
```

File: docs/usage/templates.md
Comment (HonkingGoose) [inline docs/usage/templates.md:95]:
```suggestion
If you're self-hosting Renovate, you can expose additional variables with the  [`customEnvVariables`](https://docs.renovatebot.com/self-hosted-configuration/#customenvvariables) config option.
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
First let's improve the current text.

It looks we limit the environment variables on the GitHub hosted version to improve the security, and to prevent leaking secret stuff via the environment variables.

It feels to me we should warn a lot louder that this is potentially dangerous? How do you maintainers feel about a warning admonition like this:

> **Warning**
> Think hard before you allow more environment variables.
> We recommend you stick with the default settings, unless you really need the flexibility of extra environment variables.
> Malicious scripts may leak secret information stored in environment variables.

**Comment by chopeen** [comment]
Thank you for the edits, @HonkingGoose! I applied them all.

I will wait for the maintainer to decide if the proposed warning admonition should be added.

File: lib/util/template/index.ts
Comment (rarkins) [inline lib/util/template/index.ts:215]: Are you sure we need to add every key within `env` to the allowed fields list? I was expecting that we just need to add `env` itself

Comment (chopeen) [reply lib/util/template/index.ts:215]: When the variable is `env.HOME`, the loop actually checks both the `env` and `HOME` strings, because extracted template variables are split with `match[1].split('.')`:

```
> match[1]
"env.HOME"

> const varNames = match[1].split('.');
["env", "HOME"]
```

So actually an info gets logged for `env` because it is not in `allowedFields`; for `HOME`, it's fine. This is a bug I should fix I think. :bug: 

List `allowedFieldsList` is used when `filterFields == true`, but no actual filtering happens, only messages 'Disallowed variable name in template' are getting logged - is that on purpose? :thinking: 

---

Coming back to your question, I see two solutions:

1. I can add an extra check to see of `varNames[0] == 'env'`;
2. I can add string `env` to list `allowedFields`.

Which one do you think is better?

Comment (rarkins) [reply lib/util/template/index.ts]: We should allow everything under `env`.

The log message is intentional - the goal was to make it a "soft" warning until we get all the fields in use and only later make it a hard blocking.

Comment (viceice) [reply lib/util/template/index.ts]: we already hard block that
https://github.com/renovatebot/renovate/blob/9f0a335fba39f4fb809105dd0ae187ea449ebf6f/lib/util/template/index.ts#L172-L174

Comment (chopeen) [reply lib/util/template/index.ts]: > We should allow everything under `env`.

I believe [files#diff-29b93fR217](https://github.com/renovatebot/renovate/pull/19301/files#diff-29b93fd71dde6f977077479154849d32614311d877dd83dbf1f6903401ee8088R217) now allows everything under `env`.

File: docs/usage/templates.md
Comment (viceice) [inline docs/usage/templates.md:100]: @HonkingGoose @rarkins should we add a bigger warning here?

File: lib/util/template/index.ts
Comment (viceice) [inline lib/util/template/index.ts:209]:
```suggestion
  const env = getChildEnv({});
```
should never return null or undefined. fix test mock instead

Comment (chopeen) [reply lib/util/template/index.ts]: I cannot recall why I added it in the first place.

I removed it as suggested, ran the tests and all passed.

File: docs/usage/templates.md
Comment (HonkingGoose) [reply docs/usage/templates.md:100]: We should use a warning admonition to grab the readers attention.

I think the key points to cover are:

- Explain what's the benefit of enabling the dangerous option
- Explain why it can be dangerous
- If there's any way to make it less dangerous, explain how
- Give our recommendations as maintainers
- Warn loudly about dangerous stuff

Comment (viceice) [reply docs/usage/templates.md:100]: can you make a suggestion?

**Review by viceice — COMMENTED** [review]
see open discussion

File: docs/usage/templates.md
Comment (rarkins) [reply docs/usage/templates.md:100]: I think exposeAllEnv should have the big warning while this docs section just references it

Comment (chopeen) [reply docs/usage/templates.md]: I added the big warning to `docs/usage/self-hosted-configuration.md` and left only a short remark in `docs/usage/templates.md`.

File: docs/usage/templates.md
Comment (rarkins) [inline docs/usage/templates.md:100]:
```suggestion
You can also use the [`exposeAllEnv`](https://docs.renovatebot.com/self-hosted-configuration/#exposeallenv) config option to allow all environment variables in templates, but make sure to consider the security implications of giving the scripts unrestricted access to all variables.
```

Single line per sentence

File: lib/util/template/index.ts
Comment (rarkins) [inline lib/util/template/index.ts:175]: @zharinov does this look right?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/template/index.ts
Comment (viceice) [reply lib/util/template/index.ts:175]: was my idea to simplify things.

Comment (zharinov) [reply lib/util/template/index.ts:175]: Can something go wrong because of this structure being nested? If not, then it looks okay to me.

Comment (viceice) [reply lib/util/template/index.ts:175]: `env` should be a simple string dictionary

Comment (zharinov) [reply lib/util/template/index.ts:175]: I mean, the whole data structure allows for nested `env`. To the another hand, seems like it's being attached only on top level. Well, everything looks okay to me here.

**Review by viceice — APPROVED** [review]

**Comment by chopeen** [comment]
@viceice Can you please start the CI workflow for me?

As a first-time contributor, I am not allowed to do it myself.

**Review by rarkins — APPROVED** [review]

**Comment by rarkins** [comment]
Thanks @chopeen !

**Pull request merged**
